### ISSUE TITLE
Trilinos: Possible to run CUDA without MPI?

### ISSUE THREAD

**Issue by DABH** [trilinos/Trilinos#6006]
## Question

I'd like to use a CUDA execution space for some linear solves I'm doing, but I'd like to evaluate the performance of CPU vs. GPU on my local machine, without using MPI (in particular I get "internal compiler errors" from g++/nvcc when trying to build with MPI+CUDA, so I'd like to try a simpler setup...).  However, when I try to configure and build Trilinos with CUDA but without MPI, I get a lot of errors complaining about not finding mpi.h, which seems to imply that CUDA support requires building with MPI support.

Can you confirm whether this is the case, and whether there are any plans to support single-node CUDA without MPI?

Or if I'm wrong, do you have a simple example of a set of CMake flags which would give a Trilinos build with CUDA but without MPI?

Thanks in advance for your time!

**Comment by brian-kelley** [comment]
@DABH If you have a working MPI+CUDA config script, you should be able to disable MPI with these lines:
```
-DCMAKE_C_COMPILER=gcc \
-DCMAKE_CXX_COMPILER=nvcc_wrapper \
-DTPL_ENABLE_MPI:BOOL=OFF \
```

Here's a complete config script that I used with GCC 7.4.0 and CUDA 10.1. You'll probably have to change KOKKOS_ARCH to match your machine - [see here for list of architectures](https://github.com/kokkos/kokkos/wiki/Compiling#table-43-architecture-variables). Also if ``nvcc_wrapper`` isn't in your PATH then change CMAKE_CXX_COMPILER to the absolute path ``/path/to/trilinos/packages/kokkos/bin/nvcc_wrapper``.
```
cmake \
-DTrilinos_ENABLE_MueLu=ON \
-DTrilinos_ENABLE_Ifpack2=ON \
-DTrilinos_ENABLE_Tpetra=ON \
-DTrilinos_ENABLE_Xpetra=ON \
-DTrilinos_ENABLE_KokkosKernels=ON \
-DTrilinos_ENABLE_TESTS:BOOL=ON \
-DCMAKE_BUILD_TYPE:STRING=Debug \
-DBUILD_SHARED_LIBS:BOOL=ON \
-DCMAKE_C_COMPILER=gcc \
-DCMAKE_CXX_COMPILER=nvcc_wrapper \
-DTrilinos_ENABLE_CXX11:BOOL=ON \
-DTrilinos_CXX11_FLAGS:STRING="-std=c++11 -expt-extended-lambda" \
-DCMAKE_CXX_FLAGS:STRING='-Wall -Wno-unknown-pragmas -Wno-unused-but-set-variable -Wno-inline -Wshadow' \
-DTrilinos_ENABLE_Fortran:BOOL=OFF \
-DTPL_ENABLE_CUDA:BOOL=ON \
-DTPL_ENABLE_CUSPARSE:BOOL=OFF \
-DTPL_ENABLE_HWLOC:BOOL=OFF \
-DTPL_ENABLE_BLAS:BOOL=ON \
-DTPL_ENABLE_LAPACK:BOOL=ON \
-DKokkos_ENABLE_Cuda:BOOL=ON \
-DKokkos_ENABLE_Cuda_UVM:BOOL=ON \
-DKokkos_ENABLE_Cuda_Lambda:BOOL=ON \
-DTrilinos_ENABLE_Epetra:BOOL=OFF \
-DTrilinos_ENABLE_ShyLU_Node:BOOL=OFF \
-DCMAKE_VERBOSE_MAKEFILE:BOOL=OFF \
-DTPL_ENABLE_MPI:BOOL=OFF \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
-DTpetra_INST_INT_LONG_LONG:BOOL=OFF \
-DTrilinos_ENABLE_COMPLEX:BOOL=OFF \
-DTeuchos_ENABLE_COMPLEX:BOOL=OFF \
-DTpetra_INST_COMPLEX_DOUBLE:BOOL=OFF \
-DTpetra_INST_COMPLEX_FLOAT:BOOL=OFF \
-DTpetra_INST_SERIAL=ON \
-DMueLu_ENABLE_Experimental:BOOL=ON \
-DMueLu_ENABLE_Kokkos_Refactor:BOOL=ON \
-DMueLu_ENABLE_Kokkos_Refactor_Use_By_Default:BOOL=ON \
-DXpetra_ENABLE_Experimental:BOOL=ON \
-DXpetra_ENABLE_Kokkos_Refactor:BOOL=ON \
-DKOKKOS_ENABLE_DEPRECATED_CODE=OFF \
-DKOKKOS_ARCH="SKX;Pascal61" \
../Trilinos
```

**Comment by brian-kelley** [comment]
@DABH btw, I was getting internal compiler errors too with CUDA+MPI, when I had GCC 7.2.0 instead (the generally recommended version to go with CUDA 10). Were your messages something like ``internal compiler error: in maybe_undo_parenthesized_ref`` in a Kokkos file? I had to update GCC to 7.4.0 to make that go away.

**Comment by DABH** [comment]
@brian-kelley yes that is the exact kind of error I was seeing!  Although my default `g++` version is 8.3.0 it seems like `mpicxx` or `nvcc_wrapper` is preferring some old version like 6.3.0?  Your comment gives me hope that if I can figure out how to force those compilers to use the new versions, I can get MPI+CUDA working... will mess around a bit more to see if I can get that to happen (let me know if you have any tips)... thanks!

Meanwhile I'll try your config to disable MPI as well -- thanks again

**Comment by brian-kelley** [comment]
@DABH (I'm assuming you have OpenMPI) - if $OMPI_CXX is non-empty then mpicxx will use that as the underlying compiler, instead of whatever it was built with. Although, if you using MPI+CUDA before then OMPI_CXX might point to nvcc_wrapper.

Likewise if $NVCC_WRAPPER_DEFAULT_COMPILER is non empty then nvcc_wrapper will use that as the underlying host compiler. Otherwise it should just be using "g++" from your $PATH.

**Comment by DABH** [comment]
@brian-kelley Thanks, that's helpful (yep, I'm using OpenMPI and am pointing $OMPI_CXX at nvcc_wrapper).

It looks like I was able to get a build of Trilinos with CUDA but no MPI; however, I get the following error when building my application:
```
/home/foo/trilinos-install-cuda/include/Tpetra_withLocalAccess.hpp(566): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of "void Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::withLocalAccess(Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::current_user_function_type, FirstLocalAccessType, Rest...) [with FirstLocalAccessType=Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Rest=<>]" 
(565): here
            instantiation of "void Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::withLocalAccess(Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::current_user_function_type, FirstLocalAccessType, Rest...) [with FirstLocalAccessType=Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Rest=<Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>>]" 
(565): here
            instantiation of "void Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::withLocalAccess(Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::current_user_function_type, FirstLocalAccessType, Rest...) [with FirstLocalAccessType=Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Rest=<Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>>]" 
(565): here
            instantiation of "void Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::withLocalAccess(Tpetra::Details::WithLocalAccess<FirstLocalAccessType, Rest...>::current_user_function_type, FirstLocalAccessType, Rest...) [with FirstLocalAccessType=Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::WriteOnly>, Rest=<Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>>]" 
(617): here
            instantiation of "void Tpetra::withLocalAccess(Tpetra::Details::ArgsToFunction<LocalAccessTypes...>::type, LocalAccessTypes...) [with LocalAccessTypes=<Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::WriteOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>, Tpetra::Details::LocalAccess<Tpetra::Vector<Tpetra::Details::DefaultTypes::scalar_type, Tpetra::Details::DefaultTypes::local_ordinal_type, Tpetra::Details::DefaultTypes::global_ordinal_type, Tpetra::Details::DefaultTypes::node_type>, Kokkos::CudaUVMSpace, Tpetra::Details::AccessMode::ReadOnly>>]" 
/home/foo/trilinos-install-cuda/include/Ifpack2_Details_ScaledDampedResidual_def.hpp(473): here
            [ 20 instantiation contexts not shown ]
```

I haven't seen this issue when doing a build of Trilinos using OpenMP, for instance.  I'm on `master` of Trilinos.  Any ideas what could be going on here?

**Comment by DABH** [comment]
(Meanwhile, I get a ton of warnings like this -- any idea on whether this is something I'm doing wrong, or an issue in the library?  Have you seen anything similar?)
```
/home/foo/trilinos-install-cuda/include/Tpetra_Util.hpp(655): warning: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during:
            instantiation of "void Tpetra::merge2(IT1 &, IT2 &, IT1, IT1, IT2, IT2) [with IT1=__gnu_cxx::__normal_iterator<long long *, std::vector<long long, std::allocator<long long>>>, IT2=__gnu_cxx::__normal_iterator<double *, std::vector<double, std::allocator<double>>>]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(4781): here
            instantiation of "void Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::globalAssemble() [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(5039): here
            instantiation of "void Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::fillComplete(const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, const Teuchos::RCP<Teuchos::ParameterList> &) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(8613): here
            instantiation of "Teuchos::RCP<Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>> Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::add(const Scalar &, const Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> &, const Scalar &, const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, const Teuchos::RCP<Teuchos::ParameterList> &) const [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(211): here
            instantiation of class "Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(211): here
            instantiation of "Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::CrsMatrix(const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, const Teuchos::ArrayView<const size_t> &, Tpetra::ProfileType, const Teuchos::RCP<Teuchos::ParameterList> &) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_RowMatrix_def.hpp(205): here
            instantiation of "Teuchos::RCP<Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>> Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::add(const Scalar &, const Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> &, const Scalar &, const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, const Teuchos::RCP<Teuchos::ParameterList> &) const [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(155): here
            implicit generation of "Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::RowMatrix() [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(155): here
            instantiation of class "Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Tpetra_CrsMatrix_def.hpp(155): here
            instantiation of "Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::CrsMatrix(const Teuchos::RCP<const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::map_type> &, size_t, Tpetra::ProfileType, const Teuchos::RCP<Teuchos::ParameterList> &) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Xpetra_TpetraCrsMatrix_def.hpp(55): here
            instantiation of "Xpetra::TpetraCrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::TpetraCrsMatrix(const Teuchos::RCP<const Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, size_t, Xpetra::ProfileType, const Teuchos::RCP<Teuchos::ParameterList> &) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=long long, Node=Xpetra::EpetraNode]" 
/home/foo/trilinos-install-cuda/include/Xpetra_CrsMatrixFactory.hpp(500): here
```

**Comment by brian-kelley** [comment]
@DABH I tried to replicate your Ifpack2 build error (CUDA, no MPI) but it was built successfully. Maybe somebody else from @trilinos/tpetra or @trilinos/ifpack2 can see if this is a defect in Tpetra::withLocalAccess, or the way it's used in Ifpack2_Details_Chebyshev?

As for the __host__ __device__ warnings, you can ignore them. It's a by-product of the __device__ annotation that Kokkos puts on all kernel functions, even ones only intended to be run on host.

**Comment by jhux2** [comment]
I know this works on internal Sandia clusters, so you might consider rolling back.

```
gcc/7.3.1
cuda/9.2.148
```

**Comment by DABH** [comment]
Cool, I'm using the same CUDA version but I'll give it a whirl with the older gcc version and report back!

**Comment by DABH** [comment]
Great news!  With g++ 7.4.0 instead of 8.3.0 (same CUDA version in either case) I was able to compile my application!  No code changes in my application, but I did try a couple other things to more closely replicate @brian-kelley 's config, in particular switching from C++14 to C++11, and mostly copying his configure script instead of using the one I was trying before (maybe there is some subtle difference in flags that causes the issue I was seeing).  Happy to play around with this a bit more to see if I can narrow down what really triggers this error, but meanwhile I'm satisfied to have a running program :)
